Node-wide throughput exemptions for clients

dlex opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR introduces a feature to make client_ids specified by a value, regex, or by a special selector for missing client_ids exempt from throughput control. Connections presenting a matching client_id value will not have the size of their requests accounted for throughput control, not they will be throttled.

The new configuration is forward looking to support the fully functional throughput control groups, this is why it may look too bulky for this specific PR.

Fixes #10575

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

### Features

* A client_id, or a group of client_ids specified with a regex, can now be excluded from node wide throughput control. The new cluster property `kafka_throughput_control` can be used to define throughput control groups for which Kafka traffic will not be limited by the values specified by `kafka_throughput_limit_node_*_bps`.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

dlex commented:
Force push: fixed a followup from #10285, avoid possible segfault when "kafka_throughput_controlled_api_keys" changes on the fly.

dlex commented:
Force push:
* example added to the configuration so that `test_valid_settings` does not fail on it
* use the existing `property_type_name` instead of the new memeber name to define configuration type
* validation function now returns `std::nullopt` for success as expected by `property`
* error message extended in `test_valid_settings`

dlex commented:
Force push: unit test fixed

dlex commented:
CI failure in https://buildkite.com/redpanda/redpanda/builds/29432#01883193-77a7-4f46-bac1-be8fac94ff52
* Exisinng: #10136

dlex commented:
CI Failures in https://buildkite.com/redpanda/redpanda/builds/29432#01883193-77ab-4371-bfc8-daeb63c0840b:
* existing: #10024
* existing: #10674

dlex commented:
/ci-repeat 3

dotnwat commented on src/v/config/throughput_control_group.cc:
```diff
@@ -51,8 +55,30 @@ constexpr const char* tp_limit_node_in = "throughput_limit_node_in_bps";
+};
+
+struct client_id_matcher_type {
+    struct unspecified_tag {};
+    std::variant<unspecified_tag, copyable_RE2> v;
```
is this what std::monostate is for, or would this be better represented as `optional<RE2>`?

dotnwat replied:
heh yeh, `optional<optional<re2>>` is sus, but generally i think the outcome is better when stuff is represented as directly as possible, even if it means weird stuff like nested optionals.

dotnwat commented on src/v/config/throughput_control_group.cc:
```diff
@@ -82,20 +108,36 @@ operator<<(std::ostream& os, const throughput_control_group& tcg) {
+                always_false_v<T>, "non-exhaustive client_id_re_type visitor");
+          }
+      },
+      client_id_matcher->v);
```
since this PR introduces throughput_control_group.cc why have this other commit that rewrites a bunch of it? seems like this should be squashed with the second commit to avoid extra reviewing?

dlex replied:
I was trying to split the large file into smaller parts to make reviewing easier :) If that's the other way around actually, I will squash

dotnwat replied:
yeh i mean i think it's fine the way it is. i only mentioned it because it seemed like enough code was changing that it wasn't necessarily making review easier. but it isn't a problem either way.

dotnwat commented on src/v/config/throughput_control_group.cc:
```diff
@@ -82,20 +108,36 @@ operator<<(std::ostream& os, const throughput_control_group& tcg) {
     }
-    return re2::RE2::FullMatch(
-      re2::StringPiece(client_id.c_str(), client_id.length()), *client_id_re);
+    return std::visit(
+      [client_id](auto&& v) -> bool {
+          using T = std::decay_t<decltype(v)>;
+          if constexpr (std::is_same_v<
+                          T,
+                          client_id_matcher_type::unspecified_tag>) {
+              // only missing client_id matches the unspecified tag
+              return !client_id;
+          } else if constexpr (std::is_same_v<T, copyable_RE2>) {
+              // missing client_id never matches a re
+              return client_id
+                     && re2::RE2::FullMatch(re2::StringPiece(*client_id), v);
+          } else {
+              static_assert(
+                always_false_v<T>, "non-exhaustive client_id_re_type visitor");
+          }
+      },
```
this could be `if std::holds_alternative` instead of using visit with constexpr etc... but maybe using optional woudl be even simpler?

dlex replied:
I've compared the above code with `holds_alternative` approach:
```
    if (std::holds_alternative<client_id_matcher_type::unspecified_tag>(
          client_id_matcher->v)) {
        // only missing client_id matches the unspecified tag
        return !client_id;
    }
    if (auto* const v = std::get_if<copyable_RE2>(&client_id_matcher->v); v) {
        // missing client_id never matches a re
        return client_id
               && re2::RE2::FullMatch(re2::StringPiece(*client_id), *v);
    }
    assert(false, "non-exhaustive client_id_re_type visitor");
```

So it's a comparable amount of code, a bit simpler, but it has an important drawback: a runtime fallback case. With `visit`, a missed case will cause compilation to fail, but with `holds_alternative` it will be a runtime assert/soft assert.

dlex replied:
> maybe using optional woudl be even simpler?

+1

dotnwat commented on src/v/config/throughput_control_group.cc:
```diff
@@ -106,57 +148,118 @@ ss::sstring throughput_control_group::validate() const {
+                    "{}{}", selector_prefix, selector_empty);
+              } else if constexpr (std::is_same_v<T, copyable_RE2>) {
+                  client_id_node = v.pattern();
+              } else {
+                  static_assert(
+                    always_false_v<T>,
+                    "non-exhaustive client_id_re_type visitor");
+              }
```
if you stick with the std::variant, doesn't std::visit allow you to provide overrides for each type in the variant so that you don't have to do the type inspection with constexpr and have this static_assert(false)?

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -246,38 +248,91 @@ snc_quota_manager::calc_node_quota_default() const {
     return default_quota;
 }
 
+void snc_quota_manager::get_or_create_quota_context(
+  std::unique_ptr<snc_quota_context>& ctx,
+  std::optional<std::string_view> client_id) {
+    if (likely(ctx)) {
+        // NB: comparing string_view to sstring might be suboptimal
+        if (likely(ctx->_client_id == client_id)) {
+            // the context is the right one
+            return;
```
nit: if you  made the signature `std::unique_ptr<ctx> get_or_create(std::unique_ptr<ctx>)` then we could get rid of the input/output parameter and the call site would clearly be identified at what is constructing the context. but no big deal either way.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -246,38 +248,91 @@ snc_quota_manager::calc_node_quota_default() const {
+        ctx->_exempt = true;
+        vlog(
+          klog.info,
+          "qm - Assigned throghput control group: {}",
```
spelling

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -246,38 +248,91 @@ snc_quota_manager::calc_node_quota_default() const {
+      client_id);
+    if (tcgroup_it == _kafka_throughput_control().cend()) {
+        ctx->_exempt = false;
+        vlog(klog.info, "qm - No throghput control group assigned");
```
should these be at `info` level?

dlex replied:
In the normal course that would happen once per connection lifetime, so I think info is fine. Unless we have supported clients that open and drop kafka connections all the time.

dotnwat replied:
do we log anything else at connection establishment at info level? i don't think we do...

dlex replied:
we log at INFO at connection loss [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/net/server.cc#L154-L159)

dotnwat commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -58,6 +61,29 @@ class snc_quotas_probe {
+      : _client_id(client_id) {}
+
+private:
+    friend class snc_quota_manager;
```
nit: it sort of looks like you could move snc_quota_context into snc_quota_manager.cc and then out here only deal with opaque `std::unique_ptr<ctx>` becaues it looks like all the call sites etc.. only ever pass in references to the manager itself.

dlex replied:
Indeed, I have missed that. Thanks!

dlex replied:
Having had a closer look, this makes things worse actually. Exposing unique_ptr to an incomplete class and giving away ownership of these objects requires custom deleter, which has to be complete at the point of unique_ptr declaration. If I place the deleter into `snc_quota_manager.h`, this header needs to be included into `connection_context.h` -> header bloat; putting it into a separate file is an overkill. With the current approach though, `snc_quota_manager.h` is not included into any header file, so I suppose the contention is pretty good.

BenPope replied:
You can usually forward declare the class that the `unique_ptr` holds, and then declare a destructor for the class that holds the `unique_ptr`, and the in the implementation file, you can define the destructor as default.

dlex replied:
Yeah that's a good pattern for a regular pimpl, but this case is not a pimpl actually. It's like this:

`inner.h`:
```c++
class inner_context;

class inner {
  unique_ptr<inner_context> _context;
public:
  unique_ptr<inner_context> get_context();
  ~inner();
};
```
`inner.cc`:
```c++
class inner_context {};
inner::~inner() = default;
```
`outer.h`:
```c++
class outer {
  unique_ptr<inner> _inner;
  unique_ptr<inner_context> _inner_context;
public:
  void f () {
     _inner_context = _inner->get_context();
  }
  ~outer();
};
```

For that to work, we need either of:
* `inner_context` to be complete in both `inner.cc` and `outer.cc` (where dtors for the corresponding classes are defined)
* attach a custom deleter to the `unique_ptr<inner,inner_deleter>`, and the deleter must be complete at all declarations of `unique_ptr<inner,inner_deleter>`, i.e. in both `inner.h` and `outer.h`.

The first option is apparently cleaner, causes less header dependencies, and simpler. It only needs `inner_context` to be defined in `inner.h`, and there is not much to hide there actually.

dotnwat reviewed (commented):
looks great. just a few suggestions about simplifying some stuff like the std::variant usage but those aren't blockers or anything.

dlex commented:
Force push:
* dropped `variant` in favour of `optional`
* hid `snc_quota_context`
* spelling

dlex commented:
Force push: rebased upstream

dlex commented:
Force push:
- `snc_quota_manager_context` returned to the header

dotnwat commented on src/v/kafka/server/connection_context.cc:
```diff
@@ -330,7 +357,18 @@ connection_context::dispatch_method_once(request_header hdr, size_t size) {
           }
           if (_kafka_throughput_controlled_api_keys().at(hdr.key)) {
-              _server.snc_quota_mgr().record_request_intake(size);
+              // Normally we can only get here after a prior call to
+              // snc_quota_mgr().get_or_create_quota_context() in
+              // record_tp_and_calculate_throttle(), but there is possibility
+              // that the changing configuration could still take us into this
+              // branch with unmatching (and even null) _snc_quota_context.
+              // Simply an unmatching _snc_quota_context is no big deal because
+              // it is a one off event, but we need protection from it being
+              // nullptr
```
+1

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -248,38 +249,91 @@ snc_quota_manager::calc_node_quota_default() const {
+      client_id);
+    if (tcgroup_it == _kafka_throughput_control().cend()) {
+        ctx->_exempt = false;
+        vlog(klog.info, "qm - No throughput control group assigned");
+    } else {
+        ctx->_exempt = true;
+        vlog(
+          klog.info,
```
info or debug? seems like this would be called at least once per opened connection?

dlex replied:
demoted to debug in the last update

dotnwat reviewed (commented):
everything looks good to me.

this PR adds new config options, so I expected to see at least some ducktape test that exercise some new functionality. is that missing?

dlex commented:
Two force pushes.
1. review f/ups
   * log messages on connections demoted to debug
   * stringization of `client_id_matcher_type` fixed (used to log address instead of contents)
   * ducktape test for the configuration added
2. rebased upstream

dlex commented:
> this PR adds new config options, so I expected to see at least some ducktape test that exercise some new functionality. is that missing?

@dotnwat although most configuration stuff was tested in the unit test, I have added a ducktape test - and fixed a minor bug found with it

dlex commented:
Force push: linter errors

dlex commented:
Ci failures in https://buildkite.com/redpanda/redpanda/builds/30045#01885b1e-db12-450d-bee9-e7fc3f71eac7:
* existing: #11042

dlex commented:
CI failures in https://buildkite.com/redpanda/redpanda/builds/30045#01885b1e-db19-4c8b-81d5-cf6ba965e5a0:
* existing: #10502

dlex commented:
CI failures in https://buildkite.com/redpanda/redpanda/builds/30045#01886442-4e6d-4328-b0bf-5656d157f15b:
* existing #10502
* existing #11044

dlex commented:
Force push: a typo fixed in property description

dlex commented:
Force push: rebased upstream

dotnwat commented:
/ci-repeat 5

dotnwat commented:
Failures
* https://github.com/redpanda-data/redpanda/issues/11044
* https://github.com/redpanda-data/redpanda/issues/10497
* https://github.com/redpanda-data/redpanda/issues/11166
* https://github.com/redpanda-data/redpanda/issues/8217
* https://github.com/redpanda-data/redpanda/issues/9052
* https://github.com/redpanda-data/redpanda/issues/11186
* https://github.com/redpanda-data/redpanda/issues/8496

vbotbuildovich commented:
/backport v23.1.x

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the commands below:

```
git checkout -b backport-pr-10755-v23.1.x-706 remotes/upstream/v23.1.x
git cherry-pick -x 16a6cd0bd56bb16b3aa31a8b1577d97c10655a07 a1cc8f7b9d6d68e5a3efe1dc09b615cd61a82560 bd43f4727a9887bad52d8b56c77934e755c69a67 b1fa02e1b41bc76ba57409c6106b295a67fcbdcc 7af5c60a2836e65233c433c9e068faf7c1ea302c 2ded90fd6d1ed1b7f05a07c7353a190684386844 cd9b9518bdf395e695fbedba52bca74e6b523b0e edd1f11ebd671b97b4d36f3c35e159577433a3b5 529cbf39c14e32fca87352063e951f8f39cc30b5
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/5165603381).
